### PR TITLE
(feat) Starting a visit with a past date should confirm to be retrospective visit

### PR DESCRIPTION
**Pull request by donaldkibet**
## Requirements

- [ ] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [ ] My work includes tests or is validated by existing tests.

## Summary
This PR adds ability to create a retrospective visit and load it to enable users capture data. This is a missing functionality from where users are able to load or edit past visit in retrospective visits. 

## Screenshots
Peek 2023-10-12 01-30

### CODE REVIEW

**Comment by vasharma05** [comment]
Please correct me if I'm wrong, but can't we also put an end date to a retrospective visit?

Also, see this ticket here: 
https://issues.openmrs.org/browse/O3-2180
https://issues.openmrs.org/browse/O3-2300
https://issues.openmrs.org/browse/O3-2302

Thanks!

**Comment by donaldkibet** [comment]
> Please correct me if I'm wrong, but can't we also put an end date to a retrospective visit?
> 
> Also, see this ticket here: https://issues.openmrs.org/browse/O3-2180 https://issues.openmrs.org/browse/O3-2300 https://issues.openmrs.org/browse/O3-2302
> 
> Thanks!

Thanks @vasharma05 for pointing out the tickets. 
What this PR does is that it checks if the visit start date is in the past. If that is the case. Alert the user through the modal and also capture the visit stop date in the modal.

**Comment by vasharma05** [comment]
I guess we can change the title of the PR to be more specific: "Starting a visit with a past date should confirm to be retrospective visit", or something better than this.

File: packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx [hunk 1]
```diff
@@ -140,30 +141,51 @@ const StartVisitForm: React.FC<DefaultWorkspaceProps> = ({ patientUuid, closeWor
...
+
   const onSubmit = useCallback(
     (data: VisitFormData, event) => {
-      const { timeFormat, visitDate, visitLocation, visitTime, visitType, visitAttributes } = data;
-
-      setIsSubmitting(true);
+      const { visitDate, visitLocation } = data;
 
-      const [hours, minutes] = convertTime12to24(visitTime, timeFormat);
+      const payload = createPayload(data);
```
Comment (vasharma05) [inline packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx:186]: We can use `useMemo` here

**Comment by vasharma05** [comment]
image


Why is this necessary?

**Comment by vasharma05** [comment]
> image
> Why is this necessary @donaldkibet ?

If this is necessary, we can make the max value of the calendar to today's date.
image

**Comment by vasharma05** [comment]
Things I found:
1. The running retrospective visit is not stored in the server, I mean, for the same patient I started a retrospective visit, I didn't get the retrospective visit banner on other browser.
**2. Form entry app throwing error on opening a form**
image
image

Patient: Joshua Johnson
@donaldkibet , can you please look into this?

File: packages/esm-form-entry-app/src/app/form-creation/form-creation.service.ts [hunk 2]
```diff
@@ -243,9 +243,10 @@ export class FormCreationService {
 
   private setDefaultValues(form: Form, createFormParams: CreateFormParams) {
     const { session } = createFormParams;
-
+    const currentVisitIsRetrospective = this.singleSpaPropsService.getPropOrThrow('currentVisitIsRetrospective');
+    const visitStartDatetime = this.singleSpaPropsService.getPropOrThrow('visitStartDatetime');
```
Comment (ibacher) [inline packages/esm-form-entry-app/src/app/form-creation/form-creation.service.ts:247]: Seems wrong to require this property if we only use it when a different property is set...

File: packages/esm-form-entry-app/src/app/form-creation/form-creation.service.ts [hunk 3]
```diff
@@ -243,9 +243,10 @@ export class FormCreationService {
 
   private setDefaultValues(form: Form, createFormParams: CreateFormParams) {
     const { session } = createFormParams;
-
+    const currentVisitIsRetrospective = this.singleSpaPropsService.getPropOrThrow('currentVisitIsRetrospective');
```
Comment (ibacher) [inline packages/esm-form-entry-app/src/app/form-creation/form-creation.service.ts:246]: I think here we could just assume the default of `false` right?

File: packages/esm-form-entry-app/src/app/form-creation/form-creation.service.ts [hunk 4]
```diff
@@ -243,9 +243,10 @@ export class FormCreationService {
...
+    const visitStartDatetime = this.singleSpaPropsService.getPropOrThrow('visitStartDatetime');
     // Encounter date and time.
-    const currentDate = moment().format();
+    const currentDate = currentVisitIsRetrospective ? visitStartDatetime : moment().format();
```
Comment (ibacher) [inline packages/esm-form-entry-app/src/app/form-creation/form-creation.service.ts:249]: We should probably have some sort of typecheck on `visitStartDatetime`, right?

File: packages/esm-form-entry-app/src/single-spa-props.ts [hunk 5]
```diff
@@ -3,23 +3,38 @@ import { AppProps } from 'single-spa';
...
+  closeWorkspace: () => void;
   showDiscardSubmitButtons?: boolean;
+  currentVisitIsRetrospective?: boolean;
 };
+
+// Patient Details
+type PatientDetails = {
+  patient: any;
+  patientUuid: string;
+  isOffline: boolean;
```
Comment (ibacher) [inline packages/esm-form-entry-app/src/single-spa-props.ts:35]: Is this really a patient detail?

File: packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx [hunk 6]
```diff
@@ -140,30 +141,51 @@ const StartVisitForm: React.FC<DefaultWorkspaceProps> = ({ patientUuid, closeWor
...
 
+  const createPayload = (data: VisitFormData) => {
+    const { timeFormat, visitDate, visitLocation, visitTime, visitType, visitAttributes } = data;
+    const [hours, minutes] = convertTime12to24(visitTime, timeFormat);
+
+    return {
+      patient: patientUuid,
+      startDatetime: toDateObjectStrict(
+        toOmrsIsoString(
+          new Date(dayjs(visitDate).year(), dayjs(visitDate).month(), dayjs(visitDate).date(), hours, minutes),
```
Comment (ibacher) [inline packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx:152]: I think seconds are meaningful here too?

File: packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx [hunk 7]
```diff
@@ -140,30 +141,51 @@ const StartVisitForm: React.FC<DefaultWorkspaceProps> = ({ patientUuid, closeWor
...
+        ),
+      ),
+      visitType: visitType,
+      location: visitLocation?.uuid,
```
Comment (ibacher) [inline packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx:156]: What happens when `visitLocation` is undefined? does this work?

File: packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx (hunk 1 above)
Comment (ibacher) [reply packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx:186]: We're already in a `useCallback()`... I think this is fine.

File: packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx [hunk 8]
```diff
@@ -279,6 +301,9 @@ const StartVisitForm: React.FC<DefaultWorkspaceProps> = ({ patientUuid, closeWor
     ],
   );
 
+  const handleFormSubmit = (data: VisitFormData, event) =>
+    dayjs(data.visitDate).isBefore(new Date(), 'day') ? handleRetrospectiveVisit(data, event) : onSubmit(data, event);
```
Comment (ibacher) [inline packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx:305]: Is `day` the right level of granularity for this? What if I'm processing paperwork from a same-day visit? It seems like we knew (at the beginning) whether this was retrospective or not and we should be able to pass that information to this point.

File: packages/esm-patient-chart-app/src/visit/visit-prompt/retrospective-dialog.component.tsx [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+const RetrospectiveDialog: React.FC<RetrospectiveDialogProp> = ({
+  closeModal,
+  visitDate,
+  payload,
+  patientUuid,
+  onCancel,
+}) => {
+  const { t } = useTranslation();
+
+  const retrospectiveFormSchema = useMemo(() => {
```
Comment (ibacher) [inline packages/esm-patient-chart-app/src/visit/visit-prompt/retrospective-dialog.component.tsx:51]: Why is this even part of the React object at all? Could we just as easily (and better) define this as a const?

File: packages/esm-patient-chart-app/src/visit/visit-prompt/retrospective-dialog.component.tsx [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+    return z.object({
+      visitStopDate: z
+        .date({ required_error: t('requiredVisitStopDate', 'Visit stop date is required') })
+        .max(dayjs().subtract(1, 'day').toDate(), {
+          message: t('visitStopDateBeforeCurrentDate', 'Visit stop date should be before current date'),
```
Comment (ibacher) [inline packages/esm-patient-chart-app/src/visit/visit-prompt/retrospective-dialog.component.tsx:56]: Again, I don't think this logic is generalisable across cases. Why can't I enter a retrospective visit that ends today?

File: packages/esm-patient-chart-app/src/visit/visit-prompt/retrospective-dialog.component.tsx [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+      timeFormat: z.enum(['PM', 'AM']),
+    });
+  }, []);
+
+  type RetrospectiveDialogForm = z.infer<typeof retrospectiveFormSchema>;
+  const methods = useForm<RetrospectiveDialogForm>({
+    mode: 'all',
+    resolver: zodResolver(retrospectiveFormSchema),
+    defaultValues: {
+      timeFormat: new Date().getHours() >= 12 ? 'PM' : 'AM',
```
Comment (ibacher) [inline packages/esm-patient-chart-app/src/visit/visit-prompt/retrospective-dialog.component.tsx:71]: This is also not generalisable logic. 1 AM is a valid time for inpatient actions to happen and we need to be able to record that.

**Review by ibacher — CHANGES_REQUESTED** [review]
Thanks @donaldkibet! I haven't gone through the whole thing, but a few suggestions.

**Comment by brandones** [comment]
Hi @donaldkibet . I think this needs to be reworked. RDE is supported behind a feature flag. Enabling the feature flag creates an "Add Past Visit" button, which is the UI that was agreed should be the flow for adding a past visit, in consultation with Grace and Paul. Strangely, however, rather than opening an "Add Past Visit" form, it opens the "No Active Visit" dialog. This is certainly not the right behavior for that button.

https://github.com/openmrs/openmrs-esm-patient-chart/assets/1031876/f9fc091f-56e6-4b66-a8a4-274f59cbaa1d

I have updated this relevant ticket to reflect what currently are the best requirements we have for this: [O3-2180](https://issues.openmrs.org/browse/O3-2180). This is the ticket you should address. All of the RDE work is captured in the epic [O3-2179](https://issues.openmrs.org/browse/O3-2179). I would recommend reviewing the tickets there.

To be transparent about the thinking: "start visit" and "add past visit" are very different workflows for a provider. A provider starting a visit is doing a very different kind of activity (seeing a patient) than a provider who is adding a past visit (copying some notes into the EMR). So these workflows should have separate buttons in the action menu.

Unfortunately, the RDE work became stalled on disagreements between PIH and other orgs. Fiona Anderson from PIH is a good contact about this. Paul and Grace were involved in design. If you can get this project unstuck you will be a hero. It is unclear whether the tickets that presently exist, which are based on the last convos we had with Grace & Paul, are going to reflect how the system will work in the future. Then again, if we keep building out this system as designed, and it works well, PIH may just make peace with it.

**Pull request closed without merging**